File: vcs/benchmarks/baselines/vs_go/function_call.go
// Function Call Overhead Baseline - Go Implementation
//
// This benchmark measures function call overhead in Go as a comparison
// for Verum's function call performance targets.
//
// Run: go build -o function_call function_call.go && ./function_call

package main

import (
	"fmt"
	"time"
)

const (
	warmupIterations    = 100000
	benchmarkIterations = 10000000
)

//go:noinline
func directAdd(a, b int64) int64 {
	return a + b
}

//go:noinline
func complexFunction(a, b, c int64) int64 {
	x := a * b
	y := b * c
	z := c * a
	return x + y + z
}

// Interface for virtual dispatch
type Computable interface {
	Compute(arg int64) int64
}

type Impl1 struct{ value int64 }
type Impl2 struct{ value int64 }
type Impl3 struct{ value int64 }

//go:noinline
func (i *Impl1) Compute(arg int64) int64 { return i.value + arg }

//go:noinline
func (i *Impl2) Compute(arg int64) int64 { return i.value * arg }

//go:noinline
func (i *Impl3) Compute(arg int64) int64 { return i.value - arg }

//go:noinline
func sink(x int64) {
	_ = x
}

func benchmarkInline() {
	var sum int64 = 0

	for i := 0; i < warmupIterations; i++ {
		sum = sum + 1
	}

	sum = 0
	start := time.Now()

	for i := 0; i < benchmarkIterations; i++ {
		sum = sum + 1
	}

	elapsed := time.Since(start)
	sink(sum)

	nsPerOp := float64(elapsed.Nanoseconds()) / float64(benchmarkIterations)
	fmt.Printf("inline_baseline:      %.2f ns/op\n", nsPerOp)
}

func benchmarkDirectCall() {
	var sum int64 = 0

	for i := 0; i < warmupIterations; i++ {
		sum = directAdd(sum, 1)
	}

	sum = 0
	start := time.Now()

	for i := 0; i < benchmarkIterations; i++ {
		sum = directAdd(sum, 1)
	}

	elapsed := time.Since(start)
	sink(sum)

	nsPerOp := float64(elapsed.Nanoseconds()) / float64(benchmarkIterations)
	fmt.Printf("direct_call:          %.2f ns/op\n", nsPerOp)
}

func benchmarkComplexCall() {
	var result int64 = 0

	for i := int64(0); i < warmupIterations; i++ {
		result = complexFunction(result, i, i+1)
	}

	result = 0
	start := time.Now()

	for i := int64(0); i < benchmarkIterations; i++ {
		result = complexFunction(result, i, i+1)
	}

	elapsed := time.Since(start)
	sink(result)

	nsPerOp := float64(elapsed.Nanoseconds()) / float64(benchmarkIterations)
	fmt.Printf("complex_call:         %.2f ns/op\n", nsPerOp)
}

func benchmarkInterfaceCallMonomorphic() {
	obj := &Impl1{value: 0}
	var iface Computable = obj
	var sum int64 = 0

	for i := 0; i < warmupIterations; i++ {
		sum = iface.Compute(1)
		obj.value = sum
	}

	obj.value = 0
	sum = 0
	start := time.Now()

	for i := 0; i < benchmarkIterations; i++ {
		sum = iface.Compute(1)
		obj.value = sum
	}

	elapsed := time.Since(start)
	sink(sum)

	nsPerOp := float64(elapsed.Nanoseconds()) / float64(benchmarkIterations)
	fmt.Printf("interface_mono:       %.2f ns/op\n", nsPerOp)
}

func benchmarkInterfaceCallPolymorphic() {
	objs := []Computable{
		&Impl1{value: 1},
		&Impl2{value: 2},
		&Impl3{value: 3},
	}
	var sum int64 = 0

	for i := 0; i < warmupIterations; i++ {
		sum += objs[i%3].Compute(1)
	}

	sum = 0
	start := time.Now()

	for i := 0; i < benchmarkIterations; i++ {
		sum += objs[i%3].Compute(1)
	}

	elapsed := time.Since(start)
	sink(sum)

	nsPerOp := float64(elapsed.Nanoseconds()) / float64(benchmarkIterations)
	fmt.Printf("interface_poly:       %.2f ns/op\n", nsPerOp)
}

func benchmarkClosureNoCapture() {
	closure := func(a, b int64) int64 { return a + b }
	var sum int64 = 0

	for i := 0; i < warmupIterations; i++ {
		sum = closure(sum, 1)
	}

	sum = 0
	start := time.Now()

	for i := 0; i < benchmarkIterations; i++ {
		sum = closure(sum, 1)
	}

	elapsed := time.Since(start)
	sink(sum)

	nsPerOp := float64(elapsed.Nanoseconds()) / float64(benchmarkIterations)
	fmt.Printf("closure_no_capture:   %.2f ns/op\n", nsPerOp)
}

func benchmarkClosureWithCapture() {
	multiplier := int64(2)
	offset := int64(10)
	closure := func(a int64) int64 { return a*multiplier + offset }
	var sum int64 = 0

	for i := int64(0); i < warmupIterations; i++ {
		sum = closure(i)
	}

	sum = 0
	start := time.Now()

	for i := int64(0); i < benchmarkIterations; i++ {
		sum = closure(i)
	}

	elapsed := time.Since(start)
	sink(sum)

	nsPerOp := float64(elapsed.Nanoseconds()) / float64(benchmarkIterations)
	fmt.Printf("closure_with_capture: %.2f ns/op\n", nsPerOp)
}

//go:noinline
func recursiveSum(n int64) int64 {
	if n <= 0 {
		return 0
	}
	return n + recursiveSum(n-1)
}

func benchmarkRecursiveCall() {
	depth := int64(10)
	iterations := benchmarkIterations / int(depth)

	for i := 0; i < warmupIterations/int(depth); i++ {
		sink(recursiveSum(depth))
	}

	start := time.Now()

	for i := 0; i < iterations; i++ {
		sink(recursiveSum(depth))
	}

	elapsed := time.Since(start)

	nsPerOp := float64(elapsed.Nanoseconds()) / float64(int64(iterations)*depth)
	fmt.Printf("recursive_depth_10:   %.2f ns/op\n", nsPerOp)
}

// Note: Go doesn't guarantee tail call optimization
//
//go:noinline
func tailSum(n, acc int64) int64 {
	if n <= 0 {
		return acc
	}
	return tailSum(n-1, acc+n)
}

func benchmarkTailRecursive() {
	depth := int64(100)
	iterations := benchmarkIterations / int(depth)

	for i := 0; i < warmupIterations/int(depth); i++ {
		sink(tailSum(depth, 0))
	}

	start := time.Now()

	for i := 0; i < iterations; i++ {
		sink(tailSum(depth, 0))
	}

	elapsed := time.Since(start)

	nsPerOp := float64(elapsed.Nanoseconds()) / float64(int64(iterations)*depth)
	fmt.Printf("tail_recursive_100:   %.2f ns/op\n", nsPerOp)
}

func main() {
	fmt.Println("=== Go Function Call Baseline ===")
	fmt.Println()

	benchmarkInline()
	benchmarkDirectCall()
	benchmarkComplexCall()
	benchmarkInterfaceCallMonomorphic()
	benchmarkInterfaceCallPolymorphic()
	benchmarkClosureNoCapture()
	benchmarkClosureWithCapture()
	benchmarkRecursiveCall()
	benchmarkTailRecursive()

	fmt.Println()
	fmt.Println("Verum targets (vs Go):")
	fmt.Println("  - Direct call:    Faster than Go (< 5ns vs Go's ~2-3ns)")
	fmt.Println("  - Interface call: Comparable to Go (< 15ns)")
	fmt.Println("  - Context call:   Acceptable overhead (< 30ns)")
	fmt.Println()
	fmt.Println("Note: Go has higher call overhead than C/Rust due to goroutine stack checks")
}

File: vcs/benchmarks/baselines/vs_rust/function_call.rs
//! Function Call Overhead Baseline - Rust Implementation
//!
//! This benchmark measures function call overhead in Rust as a comparison
//! for Verum's function call performance targets.
//!
//! Compile: rustc -O -o function_call function_call.rs

use std::hint::black_box;
use std::time::Instant;

const WARMUP_ITERATIONS: usize = 100_000;
const BENCHMARK_ITERATIONS: usize = 10_000_000;

#[inline(never)]
fn direct_add(a: i64, b: i64) -> i64 {
    a + b
}

#[inline(never)]
fn complex_function(a: i64, b: i64, c: i64) -> i64 {
    let x = a * b;
    let y = b * c;
    let z = c * a;
    x + y + z
}

// Trait for virtual dispatch
trait Computable {
    fn compute(&self, arg: i64) -> i64;
}

struct Impl1 { value: i64 }
struct Impl2 { value: i64 }
struct Impl3 { value: i64 }

impl Computable for Impl1 {
    #[inline(never)]
    fn compute(&self, arg: i64) -> i64 { self.value + arg }
}

impl Computable for Impl2 {
    #[inline(never)]
    fn compute(&self, arg: i64) -> i64 { self.value * arg }
}

impl Computable for Impl3 {
    #[inline(never)]
    fn compute(&self, arg: i64) -> i64 { self.value - arg }
}

fn benchmark_inline() {
    let mut sum: i64 = 0;

    for _ in 0..WARMUP_ITERATIONS {
        sum = black_box(sum + 1);
    }

    sum = 0;
    let start = Instant::now();

    for _ in 0..BENCHMARK_ITERATIONS {
        sum = black_box(sum + 1);
    }

    let elapsed = start.elapsed();
    black_box(sum);

    let ns_per_op = elapsed.as_nanos() as f64 / BENCHMARK_ITERATIONS as f64;
    println!("inline_baseline:      {:.2} ns/op", ns_per_op);
}

fn benchmark_direct_call() {
    let mut sum: i64 = 0;

    for _ in 0..WARMUP_ITERATIONS {
        sum = direct_add(sum, 1);
    }

    sum = 0;
    let start = Instant::now();

    for _ in 0..BENCHMARK_ITERATIONS {
        sum = direct_add(sum, 1);
    }

    let elapsed = start.elapsed();
    black_box(sum);

    let ns_per_op = elapsed.as_nanos() as f64 / BENCHMARK_ITERATIONS as f64;
    println!("direct_call:          {:.2} ns/op", ns_per_op);
}

fn benchmark_complex_call() {
    let mut result: i64 = 0;

    for i in 0..WARMUP_ITERATIONS as i64 {
        result = complex_function(result, i, i + 1);
    }

    result = 0;
    let start = Instant::now();

    for i in 0..BENCHMARK_ITERATIONS as i64 {
        result = complex_function(result, i, i + 1);
    }

    let elapsed = start.elapsed();
    black_box(result);

    let ns_per_op = elapsed.as_nanos() as f64 / BENCHMARK_ITERATIONS as f64;
    println!("complex_call:         {:.2} ns/op", ns_per_op);
}

fn benchmark_virtual_call_monomorphic() {
    let obj = Impl1 { value: 0 };
    let dyn_obj: &dyn Computable = &obj;
    let mut sum: i64 = 0;

    for _ in 0..WARMUP_ITERATIONS {
        sum = dyn_obj.compute(1);
    }

    sum = 0;
    let start = Instant::now();

    for _ in 0..BENCHMARK_ITERATIONS {
        sum = dyn_obj.compute(1);
    }

    let elapsed = start.elapsed();
    black_box(sum);

    let ns_per_op = elapsed.as_nanos() as f64 / BENCHMARK_ITERATIONS as f64;
    println!("virtual_monomorphic:  {:.2} ns/op", ns_per_op);
}

fn benchmark_virtual_call_polymorphic() {
    let objs: [&dyn Computable; 3] = [
        &Impl1 { value: 1 },
        &Impl2 { value: 2 },
        &Impl3 { value: 3 },
    ];
    let mut sum: i64 = 0;

    for i in 0..WARMUP_ITERATIONS {
        sum = sum.wrapping_add(objs[i % 3].compute(1));
    }

    sum = 0;
    let start = Instant::now();

    for i in 0..BENCHMARK_ITERATIONS {
        sum = sum.wrapping_add(objs[i % 3].compute(1));
    }

    let elapsed = start.elapsed();
    black_box(sum);

    let ns_per_op = elapsed.as_nanos() as f64 / BENCHMARK_ITERATIONS as f64;
    println!("virtual_polymorphic:  {:.2} ns/op", ns_per_op);
}

fn benchmark_closure_no_capture() {
    let closure = |a: i64, b: i64| -> i64 { a + b };
    let mut sum: i64 = 0;

    for _ in 0..WARMUP_ITERATIONS {
        sum = closure(sum, 1);
    }

    sum = 0;
    let start = Instant::now();

    for _ in 0..BENCHMARK_ITERATIONS {
        sum = closure(sum, 1);
    }

    let elapsed = start.elapsed();
    black_box(sum);

    let ns_per_op = elapsed.as_nanos() as f64 / BENCHMARK_ITERATIONS as f64;
    println!("closure_no_capture:   {:.2} ns/op", ns_per_op);
}

fn benchmark_closure_with_capture() {
    let multiplier: i64 = 2;
    let offset: i64 = 10;
    let closure = |a: i64| -> i64 { a * multiplier + offset };
    let mut sum: i64 = 0;

    for i in 0..WARMUP_ITERATIONS as i64 {
        sum = closure(i);
    }

    sum = 0;
    let start = Instant::now();

    for i in 0..BENCHMARK_ITERATIONS as i64 {
        sum = closure(i);
    }

    let elapsed = start.elapsed();
    black_box(sum);

    let ns_per_op = elapsed.as_nanos() as f64 / BENCHMARK_ITERATIONS as f64;
    println!("closure_with_capture: {:.2} ns/op", ns_per_op);
}

fn benchmark_closure_dyn_fn() {
    let closure: &dyn Fn(i64, i64) -> i64 = &|a, b| a + b;
    let mut sum: i64 = 0;

    for _ in 0..WARMUP_ITERATIONS {
        sum = closure(sum, 1);
    }

    sum = 0;
    let start = Instant::now();

    for _ in 0..BENCHMARK_ITERATIONS {
        sum = closure(sum, 1);
    }

    let elapsed = start.elapsed();
    black_box(sum);

    let ns_per_op = elapsed.as_nanos() as f64 / BENCHMARK_ITERATIONS as f64;
    println!("closure_dyn_fn:       {:.2} ns/op", ns_per_op);
}

fn benchmark_generic_call() {
    #[inline(never)]
    fn generic_add<T: std::ops::Add<Output = T>>(a: T, b: T) -> T {
        a + b
    }

    let mut sum: i64 = 0;

    for _ in 0..WARMUP_ITERATIONS {
        sum = generic_add(sum, 1i64);
    }

    sum = 0;
    let start = Instant::now();

    for _ in 0..BENCHMARK_ITERATIONS {
        sum = generic_add(sum, 1i64);
    }

    let elapsed = start.elapsed();
    black_box(sum);

    let ns_per_op = elapsed.as_nanos() as f64 / BENCHMARK_ITERATIONS as f64;
    println!("generic_call:         {:.2} ns/op", ns_per_op);
}

#[inline(never)]
fn recursive_sum(n: i64) -> i64 {
    if n <= 0 { 0 } else { n + recursive_sum(n - 1) }
}

fn benchmark_recursive_call() {
    let depth = 10;
    let iterations = BENCHMARK_ITERATIONS / depth;

    for _ in 0..(WARMUP_ITERATIONS / depth) {
        black_box(recursive_sum(depth as i64));
    }

    let start = Instant::now();

    for _ in 0..iterations {
        black_box(recursive_sum(depth as i64));
    }

    let elapsed = start.elapsed();

    let ns_per_op = elapsed.as_nanos() as f64 / (iterations * depth) as f64;
    println!("recursive_depth_10:   {:.2} ns/op", ns_per_op);
}

#[inline(never)]
fn tail_sum(n: i64, acc: i64) -> i64 {
    if n <= 0 { acc } else { tail_sum(n - 1, acc + n) }
}

fn benchmark_tail_recursive() {
    let depth = 100;
    let iterations = BENCHMARK_ITERATIONS / depth;

    for _ in 0..(WARMUP_ITERATIONS / depth) {
        black_box(tail_sum(depth as i64, 0));
    }

    let start = Instant::now();

    for _ in 0..iterations {
        black_box(tail_sum(depth as i64, 0));
    }

    let elapsed = start.elapsed();

    let ns_per_op = elapsed.as_nanos() as f64 / (iterations * depth) as f64;
    println!("tail_recursive_100:   {:.2} ns/op", ns_per_op);
}

fn main() {
    println!("=== Rust Function Call Baseline ===\n");

    benchmark_inline();
    benchmark_direct_call();
    benchmark_complex_call();
    benchmark_virtual_call_monomorphic();
    benchmark_virtual_call_polymorphic();
    benchmark_closure_no_capture();
    benchmark_closure_with_capture();
    benchmark_closure_dyn_fn();
    benchmark_generic_call();
    benchmark_recursive_call();
    benchmark_tail_recursive();

    println!();
    println!("Verum targets (vs Rust):");
    println!("  - Direct call:  Equivalent to Rust (< 5ns)");
    println!("  - Virtual call: Equivalent to Rust dyn (< 15ns)");
    println!("  - Closure:      Equivalent to Rust Fn (< 10ns)");
    println!("  - Generic:      Equivalent (monomorphized)");
}
